pkg/cloud/services/compute/instance_types.go

```diff
@@ -132,51 +135,100 @@ func (is *InstanceStatus) InstanceIdentifier() *InstanceIdentifier {
-	// https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/4debc1fc4742e483302b0c36b16c076977bd165d/pkg/cloud/services/compute/instance.go#L973-L998
-	// XXX: Fix this
-	for _, vifs := range ns.addresses {
-		for _, vif := range vifs {
-			if vif.Version == 4.0 && vif.Type != "floating" {
-				return vif.Address
+			var addressType corev1.NodeAddressType
+			switch address.Type {
+			case "floating":
+				addressType = corev1.NodeExternalIP
+			case "fixed":
+				addressType = corev1.NodeInternalIP
+			default:
+				// Ignore unknown address types
+				continue
 			}
+
+			addresses = append(addresses, corev1.NodeAddress{
+				Type:    addressType,
+				Address: address.Address,
+			})
 		}
+
+		addressesByNetwork[networkName] = addresses
 	}
-	return ""
+
+	return &InstanceNetworkStatus{addressesByNetwork}, nil
+}
+
+// Addresses returns a list of NodeAddresses containing all addresses which will
+// be reported on the OpenStackMachine object.
+func (ns *InstanceNetworkStatus) Addresses() []corev1.NodeAddress {
+	// We want the returned order of addresses to be deterministic to make
+	// it easy to detect changes and avoid unnecessary updates. Iteration
+	// over maps is non-deterministic, so we explicitly iterate over the
+	// address map in lexical order of network names. This order is
+	// arbitrary.
+	networks := make([]string, 0, len(ns.addresses))
+	for network := range ns.addresses {
+		networks = append(networks, network)
+	}
+	sort.Strings(networks)
+
+	var addresses []corev1.NodeAddress
+	for _, network := range networks {
+		addressList := ns.addresses[network]
+		addresses = append(addresses, addressList...)
+	}
+
+	return addresses
 }
 
-func (ns *InstanceNetworkStatus) FloatingIP() string {
-	// Return the last listed floating IPv4 from the last listed network
-	// This behaviour is wrong, but consistent with the previous behaviour
-	// https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/4debc1fc4742e483302b0c36b16c076977bd165d/pkg/cloud/services/compute/instance.go#L973-L998
-	// XXX: Fix this
-	for _, vifs := range ns.addresses {
-		for _, vif := range vifs {
-			if vif.Version == 4.0 && vif.Type == "floating" {
-				return vif.Address
+func (ns *InstanceNetworkStatus) firstAddressByNetworkAndType(networkName string, addressType corev1.NodeAddressType) string {
+	if addressList, ok := ns.addresses[networkName]; ok {
+		for i := range addressList {
+			address := &addressList[i]
+			if address.Type == addressType {
+				return address.Address
 			}
 		}
 	}
 	return ""
 }
+
+// IP returns the first listed ip of an instance for the given network name.
+func (ns *InstanceNetworkStatus) IP(networkName string) string {
+	return ns.firstAddressByNetworkAndType(networkName, corev1.NodeInternalIP)
+}
+
+// FloatingIP returns the first listed floating ip of an instance for the given
+// network name.
+func (ns *InstanceNetworkStatus) FloatingIP(networkName string) string {
```

Comment:
I guess I lost some background
usually ,multiple network is reasonable, so we have multiple NICs from various network
but multiple NIC from one network is not so familiar to me, anyway, seems no harm if we 
return the first ip

Reply:
We're actually only using this method in 1 place: APIInstance() above. And the only thing that uses APIInstance() is the bastion host. I guess we're not likely to put a second floating IP on the bastion host, so this is really just something to make the behaviour defined if we ever did.